collector/mdadm.go

```diff
@@ -0,0 +1,200 @@
+// +build !nomdadm
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/log"
+)
+
+var statusfile = "/proc/mdstat"
+
+type mdstatus struct {
+	mdName       string
+	isActive     bool
+	disksActive  int64
+	disksTotal   int64
+	blocksTotal  int64
+	blocksSynced int64
+}
+
+type mdadmCollector struct {
+	isActive     *prometheus.GaugeVec
+	disksActive  *prometheus.GaugeVec
+	disksTotal   *prometheus.GaugeVec
+	blocksTotal  *prometheus.GaugeVec
+	blocksSynced *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["mdadm"] = NewMdadmCollector
+}
+
+func evalStatusline(statusline string) (active, total, size int64) {
+
+	first_split := strings.Split(statusline, "[")
+	second_split := strings.Split(first_split[1], "]")
+
+	sizeDirty := strings.Split(first_split[0], "blocks")[0]
+	sizeStripped := strings.Replace(sizeDirty, " ", "", -1)
+
+	numberOfDrives := strings.Split(second_split[0], "/")
+
+	total, _ = strconv.ParseInt(numberOfDrives[0], 10, 64)
+	active, _ = strconv.ParseInt(numberOfDrives[1], 10, 64)
+	size, _ = strconv.ParseInt(sizeStripped, 10, 64)
+
+	return active, total, size
+
+}
+
+// gets the size that has already been synced out of the sync-line
+func evalBuildline(buildline string) int64 {
+	first_split := strings.Split(buildline, "(")
+	size := strings.Split(first_split[1], "/")[0]
+
+	syncedSize, _ := strconv.ParseInt(size, 10, 64)
+
+	return syncedSize
+}
+
+// parses an mdstat-file and returns a struct with the relevant infos
+func parseMdstat(mdstatusfilepath string) ([]mdstatus, error) {
+
+	content, err := ioutil.ReadFile(mdstatusfilepath)
+	if err != nil {
+		return []mdstatus{}, err
+	}
+
+	mdstatusfile := string(content)
+
+	lines := strings.Split(mdstatusfile, "\n")
+	var currentMd string
+
+	// each md has at least the deviceline, statusline and one empty line afterwards
+	// so we will have probably something of the order len(lines)/3 devices
+	// so we use that for preallocation
+	estimateMds := len(lines) / 3
+	mdstates := make([]mdstatus, 0, estimateMds)
+
+	for i, l := range lines {
+		if !(len(l) > 0) {
+			// skip entirely empty lines
+			continue
+		}
+
+		if string(l[0]) == " " {
+			// those lines are not the beginning of a md-section
+			continue
+		} else {
+
+			if strings.HasPrefix(l, "Personalities") || strings.HasPrefix(l, "unused") {
+				// we aren't interested in lines with general info
+				continue
+			}
+
+			currentMd = strings.Split(l, " ")[0]         // name of the md-device we are parsing atm
+			isActive := !strings.Contains(l, "inactive") // activity-status of said md-device
```

Comment:
There could conceivably be a device with 'inactive' in the name, can you check the first field after the colon?